Add Quat::slerp

aloucks opened the pull request:
This PR adds `Quat::slerp`. This implementation is derived from the cgmath implementation which is based on http://number-none.com/product/Understanding%20Slerp,%20Then%20Not%20Using%20It/

Unlike the cgmath implementation , intrinsics are used here when available.

Opening as draft to gain feedback.

1. `_MM_SHUFFLE` is considerably more readable than the duplication of the binary constants. Although it's implemented in the rust std library, it's currently nightly only. I've copied that implementation here.
2. I've added a few macros that mirror the DirectXMath macros, but I'm not sure if these are something you'll want in the code.
3. I've used the DirectXMath constant names which aren't snake case. I think it makes it easier when comparing the code, but I can understand if you wouldn't want that.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -85,6 +92,28 @@ pub(crate) mod sse2 {
     // _ps_const_ty!(PS_COSCOF_P2, f32x4, 4.166_664_6e-2);
     // _ps_const_ty!(PS_CEPHES_FOPI, f32x4, 1.273_239_5); // 4 / M_PI
 
+    _ps_const_ty!(g_XMNegativeZero, u32x4, 0x80000000);
+    _ps_const_ty!(g_XMPi, f32x4, std::f32::consts::PI);
+    _ps_const_ty!(g_XMHalfPi, f32x4, std::f32::consts::FRAC_PI_2);
+    _ps_const_ty!(g_XMSinCoefficients0 , f32x4, -0.16666667, 0.0083333310, -0.00019840874, 2.7525562e-06);
+    _ps_const_ty!(g_XMSinCoefficients1 , f32x4, -2.3889859e-08, -0.16665852 /*Est1*/, 0.0083139502 /*Est2*/, -0.00018524670 /*Est3*/);
+    _ps_const_ty!(g_XMOne, f32x4, 1.0);
+    _ps_const_ty!(g_XMTwoPi, f32x4, std::f32::consts::PI * 2.0);
+    _ps_const_ty!(g_XMReciprocalTwoPi , f32x4, 0.159154943);
```
I would want these to use Rust naming conventions.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -85,6 +92,28 @@ pub(crate) mod sse2 {
+    // XMVectorNegativeMultiplySubtract
+    macro_rules! XM_FMADD_PS {
+        ($a:expr, $b:expr, $c:expr) => {
+            _mm_fmadd_ps($a, $b, $c)
```
Need to check the fma target feature is available to use this.

aloucks replied:
I've updated these detect target CPU features.

bitshifter replied:
In addition to the request to make this an inline function and rename it, there are `mul_add` methods on `Vec3` and `Vec4` that could call this when `vec3sse2` and `vec4sse2` are enabled respectively.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -85,6 +92,28 @@ pub(crate) mod sse2 {
+    _ps_const_ty!(g_XMTwoPi, f32x4, std::f32::consts::PI * 2.0);
+    _ps_const_ty!(g_XMReciprocalTwoPi , f32x4, 0.159154943);
+
+    macro_rules! XM_PERMUTE_PS {
```
I'm not sure it's worth keeping these macros. I think they can be useful when dealing with different implementations of common operations, (e.g. SSE2 and SSE4) but so far I've just stuck with SSE2 and written everything in explicit intrinsices.

aloucks replied:
The `fma` instructions seem to add a significant speedup. 

```
+avx ~5%
+fma ~20%
+fma,+avx ~20%
```

aloucks replied:
Hmm, I seem to get different results when running all of the quat tests vs just the quat slerp test. The `fma` instruction is without question an improvement. It's not clear whether the `avx` one is useful or not.

bitshifter replied:
Not sure what you mean, the quat slerp test is changing the result of the other tests? Or enabling AVX is changing results? Having FMA off or on could definitely give different results.

bitshifter replied:
With `XM_PERMUTE_PS` and `XM_FMADD_PS` I'd prefer to make them inline functions and name them `m128_permute` and `m128_mul_add`. Note that `avx` and `fma` features aren't enabled by default in x64 builds and I probably won't add build permutations for these to the CI, just because it already takes an hour as is so they won't be getting tested. It might be worth making a note about that in the README. Well, if I could add a travis build that wasn't waited on by the CI (so PR's weren't waiting on them) but still let me know if they are broken maybe I could add an automated test.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +168,69 @@ pub(crate) mod sse2 {
         _mm_or_ps(result, _mm_castsi128_ps(test))
     }
 
+    #[inline]
+    #[allow(non_snake_case)]
+    pub(crate) const fn _MM_SHUFFLE(z: u32, y: u32, x: u32, w: u32) -> i32 {
+        // core::arch::x86_64::_MM_SHUFFLE requires nightly
+        ((z << 6) | (y << 4) | (x << 2) | w) as i32
+    }
```
I prefer the binary literals. I used to use `_MM_SHUFFLE` but actually found it harder to follow what was going on so migrated to binary literals instead.

aloucks replied:
I've changed this to use binary literals

bitshifter commented on src/f32/quat.rs:
```diff
@@ -358,6 +358,64 @@ impl Quat {
+            // assumes lerp returns a normalized quaternion
+            self.lerp(end, s)
+        } else {
+            #[cfg(not(vec4sse2))]
```
use `#[cfg(vec4f32)]`, at some point I hope to support other architectures so `#[cfg(not(vec4sse2))]` would be incorrect once that happens.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +168,69 @@ pub(crate) mod sse2 {
+    }
+
+    /// From DirectXMath: XMVectorSin
+    pub(crate) unsafe fn m128_sin(v: __m128) -> __m128 {
```
Would be good to add a test for this similar to the `scalar_acos` test to get an idea of the error.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -85,6 +92,28 @@ pub(crate) mod sse2 {
+        }
+    }
+
+    // XMVectorNegativeMultiplySubtract
```
This comment looks wrong, this is fused multiply add

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +168,69 @@ pub(crate) mod sse2 {
+        ((z << 6) | (y << 4) | (x << 2) | w) as i32
+    }
+
+    /// From DirectXMath: XMVectorModAngles
```
I'd put this comment inside the function rather than the doc comment.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +168,69 @@ pub(crate) mod sse2 {
+        XM_FMADD_PS!(g_XMTwoPi.m128, v, angles) // XMVectorNegativeMultiplySubtract
+    }
+
+    /// From DirectXMath: XMVectorSin
```
Same with this comment

bitshifter commented on src/f32/quat.rs:
```diff
@@ -358,6 +358,64 @@ impl Quat {
+    /// is `1.0`, the result will be equal to `other`.
+    #[inline]
+    pub fn slerp(self, end: Self, s: f32) -> Self {
+        // http://number-none.com/product/Understanding%20Slerp,%20Then%20Not%20Using%20It/
```
I was going to add slerp when I wrote glam, read this article and decided not to :) Well, partially that was the reason. I was going for a minimal feature set which was somewhat based on what https://github.com/nfrechette/rtm supported. It also doesn't have a slerp. Are you using slerp for something?

aloucks replied:
Yes. I was looking at migrating some stuff from cgmath to glam and slerp was missing. I was checking animations with some of the [glTF sample models](https://github.com/KhronosGroup/glTF-Sample-Models) and the lack of constant velocity makes `lerp` unsuitable.

bitshifter replied:
Thanks, that's good to know.

aloucks commented:
@bitshifter I've addressed all the feedback and added some tests for slerp.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +197,64 @@ pub(crate) mod sse2 {
     }
 
+    /// Returns a vector whose components are the corresponding components of Angles modulo 2PI.
+    #[allow(dead_code)]
+    pub(crate) unsafe fn m128_mod_angles(angles: __m128) -> __m128 {
+        // From DirectXMath: XMVectorModAngles
+        let v = _mm_mul_ps(angles, XM_RECIPROCAL_TWO_PI.m128);
+        let v = m128_round(v);
+        XM_FMADD_PS!(XM_TWO_PI.m128, v, angles)
+    }
```
If this is no longer used it should probably be removed. Although I ask about whether this should be used below.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +197,64 @@ pub(crate) mod sse2 {
+        // but this adds about 25% overhead.
+        //
+        // // Force the value within the bounds of pi
+        // let mut x = m128_mod_angles(v);
```
Curious why this was commented out? From my experience with the approximation I was using here which I commented out, accuracy started to get a large error and angles grew beyond PI.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -349,3 +465,25 @@ fn test_scalar_sin_cos() {
+
+    fn test_sse2_m128_sin_angle(a: f32) {
+        let v = Vec4::splat(a);
+        let v = unsafe { Vec4(sse2::m128_sin(v.0)) };
```
Does this function return the sin calculation in each lane? Can you pass in different angles in each lane in the input and get different results in each lane? Might be worth commenting one way or another. If it is actually returning a result in each lane then the test should test all lanes.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -349,3 +465,25 @@ fn test_scalar_sin_cos() {
+        assert_approx_eq!(v.x(), a.sin(), 1e-6);
+    }
+
+    let mut a = -PI;
```
In `test_scalar_sin_cos` I'm testing a much larger range, and accuracy did turn out to be a problem for large input values. I think it would be better to copy that test, unless this sin is only being used by slerp and the input it expected to always be in the [-PI, PI] range then it might be OK to not include the `m128_mod_angles` call and not test a larger range. However it should be clearly commented that input needs to be [-PI, PI] and asserts added to check that is the case. Otherwise that `m128_mod_angles` call probably needs to got back in and this test should test across a large range of values. I guess if this is all sounding like too much work you could just use `f32:sin` and leave an optimised approximation for another day :)

aloucks commented:
`_mm_permute_ps` and `_mm_shuffle_ps` have `#[rustc_args_required_const(2)]` attribute set to enforce that the mask is constant. Using a macro is the only option.

```rust
pub(crate) unsafe fn m128_permute(a: __m128, mask: i32) -> __m128 {
    _mm_permute_ps(a, mask)
}
```

```
error: argument 2 is required to be a constant
   --> src\f32\funcs.rs:124:9
    |
124 |         _mm_permute_ps(a, mask)
    | 
```

aloucks commented:
@bitshifter I've added doc comments on `m128_sin` to explain that `sin` is calculated in each lane and that the error increases as the input drifts further from `[-PI, PI]`. 

The function is suitable for use in `slerp` as we know the input will always be within these bounds. I've also enabled the angle modulus so that it's now  implemented exactly as `XMVectorSin`. I compared this with the `C++` version of `XMVectorSin` and observed the same behavior.

I can remove the commented out test that attempts to test a larger range (and fails), if you'd like. I also checked using `f32::sin()` for each lane  and `slerp` was about twice as slow.

I have not updated vec3/vec4 to take advantage of the fused multiply add/sub. I started to do so, but I ran into an issue with importing the the macros and I didn't want to start re-organizing things in this PR.

bitshifter commented on src/f32/funcs.rs:
```diff
@@ -139,6 +197,64 @@ pub(crate) mod sse2 {
+        // 11-degree minimax approximation
+        //
+        // Force the value within the bounds of pi
+        let mut x = m128_mod_angles(v);
```
Now that you are using mod angles again, is the accuracy OK with values outside of +/- PI? Just wondering if the function comment is still accurate?

aloucks replied:
That depends on your definition of "OK". It seems that the error increases as you move further away from `[-PI, PI]` in either direction, but its still functions as a working `sin`. The output was rubbish outside the bounds of `PI` without `mod_angles`.
